File: backend/agents/publicSafetyAgent.js
import { callGemmaVision } from './gemmaClient.js';

const PUBLIC_SAFETY_PROMPT = `You are the Public Safety Agent for ConstructGuard AI.

Analyze this construction site image for risks to the PUBLIC — pedestrians, cyclists, drivers.

Respond ONLY with valid JSON:
{
  "pedestrians_detected": { "count": 0, "too_close_to_site": true|false },
  "pedestrian_risks": ["list each risk situation"],
  "vehicle_risks": ["list risks to passing cars/cyclists"],
  "barrier_status": {
    "hoarding_complete": true|false,
    "cones_adequate": true|false,
    "signage_adequate": true|false,
    "lighting_adequate": true|false
  },
  "sidewalk_status": "clear|blocked|unsafe|not_visible",
  "missing_barriers": ["describe what's missing and where"],
  "public_risk_level": "NONE|LOW|MEDIUM|HIGH|CRITICAL",
  "summary": "One sentence"
}`;

export async function runPublicSafetyAgent({ imageBase64, imageMimeType }) {
  console.log('[PublicSafetyAgent] Checking pedestrian and public risks...');
  if (!imageBase64) {
    return {
      pedestrians_detected: { count: 0 },
      pedestrian_risks: [],
      vehicle_risks: [],
      barrier_status: {},
      sidewalk_status: 'not_visible',
      missing_barriers: [],
      public_risk_level: 'NONE',
      summary: 'No image provided',
    };
  }
  return callGemmaVision(imageBase64, imageMimeType, PUBLIC_SAFETY_PROMPT);
}

File: backend_py/agents/public_safety_agent.py
from typing import Any

from models.gemma_adapter import gemma

PUBLIC_SAFETY_PROMPT = '''You are the Public Safety Agent for ConstructGuard AI.
Analyze this construction site image for risks to pedestrians, cyclists, drivers, and nearby public areas.
Respond only with valid JSON:
{
  "pedestrians_detected": { "count": 0, "too_close_to_site": false },
  "pedestrian_risks": [],
  "vehicle_risks": [],
  "barrier_status": { "hoarding_complete": false, "cones_adequate": false, "signage_adequate": false, "lighting_adequate": false },
  "sidewalk_status": "clear|blocked|unsafe|not_visible",
  "missing_barriers": [],
  "public_risk_level": "NONE|LOW|MEDIUM|HIGH|CRITICAL",
  "summary": "one sentence"
}'''


def fallback(message: str = 'Local vision model unavailable') -> dict[str, Any]:
    return {
        'pedestrians_detected': {'count': 0, 'too_close_to_site': False},
        'pedestrian_risks': [],
        'vehicle_risks': [],
        'barrier_status': {},
        'sidewalk_status': 'not_visible',
        'missing_barriers': [],
        'public_risk_level': 'NONE',
        'summary': message,
    }


def run_public_safety_agent(image_base64: str | None, image_mime_type: str | None) -> dict[str, Any]:
    if not image_base64:
        return fallback('No image provided')
    try:
        return gemma.analyze_image(image_base64, image_mime_type or 'image/jpeg', PUBLIC_SAFETY_PROMPT)
    except Exception as exc:
        data = fallback('Public safety agent used fallback output')
        data['error'] = str(exc)
        return data
